File: server/machine/component.py
class Component:
    def __init__(self, id, name):
        self.id = id
        self.name = name
        self._setting = ''
        self._setting_params = {}
        self._on = False

    @property
    def setting(self):
        return self._setting

    @setting.setter
    def setting(self, val):
        self._setting = val

    def on(self):
        self._turn_on_current_setting()

    def off(self):
        self._on = False
        self._turn_off()

    def update_params(self, params):
        self._setting_params.update(params)
        self._on = True
        self._turn_on_current_setting()

    def set_setting_and_params(self, setting, params):
        self.setting = setting
        self._setting_params.update(params)
        self._on = True
        self._turn_on_current_setting()

    def state(self):
        return {
            'id': self.id,
            'name': self.name,
            'on': self._on,
            'setting': self.setting,
            'params': self._setting_params
        }

    def _param_generator(self, name):
        while True:
            yield self._setting_params[name] if name in self._setting_params else 0

    def _turn_off(self):
        raise Exception("_turn_off not implemented")

    def _turn_on_current_setting(self):
        raise Exception("_turn_on_current_setting not implemented")

File: client/main/elements/component.js
import * as api from '../services/componentApi'

class Component {

    constructor(state, updateState) {
        this.state = {...state};
        this._updateState = updateState;
        this.lastApiParamUpdate = new Date()
    }

    turnOn = () => {
        api.on(this.state.id);
        this._updateState({on: true})
    };

    turnOff = () => {
        api.off(this.state.id);
        this._updateState({on: false})
    };

    setSetting = (name, params = {}, onSuccess, onFailure) => {
        api.setSetting(name, this.state.id, params, onSuccess, onFailure);
        this._updateState({setting: name})
    };

    updateParam(name, value) {
        this.setParam(name, value, true);
    }

    setParam(name, value, forceApiUpdate = false, apiUpdateInterval = 200) {
        this._updateState({params: {...this.state.params, [name]: value}});
        if (forceApiUpdate || new Date() - this.lastApiParamUpdate > apiUpdateInterval) {
            this.lastApiParamUpdate = new Date();
            const params = {};
            params[name] = value;
            api.updateParams(this.state.id, params);
        }
    }

    get id() {
        return this.state.id
    }

    get name() {
        return this.state.name
    }

    get setting() {
        return this.state.setting
    }

    get params() {
        return this.state.params
    }

    get on() {
        return this.state.on
    }

    get type() {
        return this.state.type
    }
}

export default Component